ffmpeg/api_test.go

```diff
@@ -478,3 +478,209 @@ func TestTranscoder_API_AlternatingTimestamps(t *testing.T) {
+	tc.StopTranscoder()
+	tc = NewTranscoder()
+	for i := 0; i < 4; i++ {
+		fname := fmt.Sprintf("%s/short%d.ts", dir, i)
+		oname := fmt.Sprintf("%s/vpassthru%d.ts", dir, i)
+		out := []TranscodeOptions{{Profile: P144p30fps16x9, Accel: accel}}
+		out[0].Profile.Framerate = 0 // Passthrough!
+
+		out[0].Oname = oname
+		in := &TranscodeOptionsIn{Fname: fname, Accel: accel}
+		res, err := tc.Transcode(in, out)
+		if err != nil {
+			t.Error("Could not transcode: ", err)
+		}
+		// verify that output frame count is same as input frame count
+		if res.Decoded.Frames != fc || res.Encoded[0].Frames != fc {
+			t.Error("Did not get expected frame count; got ", res.Encoded[0].Frames)
+		}
+	}
+
+	// test low fps (8) to low fps (2)
+	tc.StopTranscoder()
+	tc = NewTranscoder()
+
+	cmd = `
+    # convert segment to 8fps
+	frame_count=%d
+	ffmpeg -loglevel warning -i test.ts -vf fps=8/1 -c:v libx264 -c:a copy test8.ts
+
+	# create short segment with #fc frames
+    ffmpeg -loglevel warning -ss 0 -i test8.ts -c copy -frames:v $frame_count -copyts short8.ts
+
+	# sanity check
+	ffprobe -loglevel warning -show_streams short8.ts | grep r_frame_rate=8/1
+	ffprobe -loglevel warning -count_frames -show_streams -select_streams v short8.ts | grep nb_read_frames=$frame_count
+  `
+	run(fmt.Sprintf(cmd, fc))
+
+	fname := fmt.Sprintf("%s/short8.ts", dir)
+	in := &TranscodeOptionsIn{Fname: fname, Accel: accel}
+	out := []TranscodeOptions{{Oname: dir + "/out8.ts", Profile: P144p30fps16x9, Accel: accel}}
+	out[0].Profile.Framerate = 2 // Force 2fps
+	res, err := tc.Transcode(in, out)
+	if err != nil {
+		t.Error(err)
+	}
+	if fc != res.Decoded.Frames {
+		t.Error("Did not decode expected number of frames: ", res.Decoded.Frames)
+	}
+	if 0 == res.Encoded[0].Frames {
+		t.Error("Did not encode any frames: ", res.Encoded[0].Frames)
+	}
+
+	// test a bunch of weird cases together
+	tc.StopTranscoder()
+	tc = NewTranscoder()
+	profile_low_fps := P144p30fps16x9
+	profile_low_fps.Framerate = uint(fc) // use the input frame count as the output fps, why not
+	profile_passthrough_fps := P144p30fps16x9
+	profile_passthrough_fps.Framerate = 0
+	for i := 0; i < 4; i++ {
+		fname := fmt.Sprintf("%s/short%d.ts", dir, i)
+		in := &TranscodeOptionsIn{Fname: fname, Accel: accel}
+		out := []TranscodeOptions{
+			{
+				Oname:        fmt.Sprintf("%s/lowfps%d.ts", dir, i),
+				Profile:      profile_low_fps,
+				AudioEncoder: ComponentOptions{Name: "copy"},
+				Accel:        accel,
+			},
+			{
+				Oname:        fmt.Sprintf("%s/copyall%d.ts", dir, i),
+				VideoEncoder: ComponentOptions{Name: "copy"},
+				AudioEncoder: ComponentOptions{Name: "drop"},
+				Accel:        accel,
+			},
+			{
+				Oname:   fmt.Sprintf("%s/passthru%d.ts", dir, i),
+				Profile: profile_passthrough_fps,
+				Accel:   accel,
+			},
+		}
+		res, err := tc.Transcode(in, out)
+		if err != nil {
+			t.Error(err)
+		}
+		if res.Encoded[0].Frames == 0 || res.Encoded[1].Frames != 0 || res.Encoded[2].Frames != fc {
```

Comment:
Checking for `res.Encoded[0].Frames == 0 ` makes me sweat a bit, because we don't really know what might be a reasonably correct value. But I understand it's pretty hard to get predictable results across different runs, probably OK for now.